src/types.rs

```diff
@@ -93,22 +93,65 @@ pub(crate) fn least_upper_bound(mut tags: impl Iterator<Item = TypeBound>) -> Ty
     .into_inner()
 }
 
+#[derive(Clone, PartialEq, Debug, Eq, derive_more::Display, Serialize, Deserialize)]
+/// Representation of a Sum type.
+/// Either store the types of the variants, or in the special (but common) case
+/// of a "simple predicate" (sum over empty tuples), store only the size of the predicate.
+enum SumType {
+    #[display(fmt = "SimplePredicate({})", "_0")]
+    Simple(u8),
```

Comment:
Why not use `usize` here?

Reply:
The thinking being that simple predicates branching factor >255 is not worth optimising for, and u8 will allow memory and serialisation optimisations.